### DISCUSSION TITLE
How to create JSON array of array of values from table?

### DISCUSSION THREAD

**Discussion by tonne1** [duckdb/duckdb#12665] [Q&A] [answered]
Hi, I would like to generate a JSON array of array of values from a table, as part of a more complex JSON:

```sql
CREATE TABLE cities AS
SELECT *
FROM (VALUES ('Amsterdam', 1), ('London', 2)) cities(name, id);
```

The following works, but is there a more generic way?
```sql
select to_json(list([name, id::varchar])) col from cities;
```
For example, query all columns of the table without enumerating them all?
Or query some and cast them all, but not by casting each of them separately?

**Comment by pkoppstein** [comment]
Use list_transform in conjunction with the idiom `select t::json j from t`:
```
D select list_transform(json_keys(j), k -> (j->k)) as json from (select cities::json j from cities);
```

```
┌──────────────────┐
│       json       │
│      json[]      │
├──────────────────┤
│ ["Amsterdam", 1] │
│ ["London", 2]    │
└──────────────────┘
```

You can use this technique in conjunction with `EXCLUDE` and other DuckDB enhancments related to column selection; and further refinements can of course be accomplished using list_filter and/or within the context of the call to list_transform

**Comment by tonne1** [comment]
Thanks for the pointer, but the returned structure in your example is not an array of arrays, and my attempts to convert it so far failed. The query

```sql
select to_json(list([name, id::varchar])) col from cities;
```
produces this result:
```
┌────────────────────────────────────┐
│                col                 │
│                json                │
├────────────────────────────────────┤
│ [["Amsterdam","1"],["London","2"]] │
└────────────────────────────────────┘
```

Also the expression is quite complex for a basic task like this.

**Comment by pkoppstein** [comment, marked as answer]
I thought the obstacle you were seeking to overcome was how to avoid naming the columns.

Anyway, you could achieve the desired result by adding `array_agg(columns (*))::JSON`, e.g.
```
with cte as 
  (select list_transform(json_keys(j), k -> (j->k)) as json from (select cities::json j from cities))
select array_agg(COLUMNS(*))::JSON from cte;
```

Agreed, this is cumbersome, to say the least.  (Welcome to SQL!). The hard thing to swallow is that it seems one cannot abstract the above satisfactorily as macros don't seem to allow table names as formal parameters.  The best I can come up with is:
```
-- Collapse the table in `cte` to a JSON array of arrays, each of which represents a row
CREATE OR REPLACE MACRO array_of_arrays_of_cte() as TABLE
  (with _inner_ as
   (select list_transform(json_keys(j), k -> (j->k)) as json from (select cte::json j from cte))
   select array_agg(COLUMNS(*))::JSON from _inner_);
```

So now one can write:
```
with cte as (from cities)
 from array_of_arrays_of_cte();
```

Extending your example slightly, we find:
```
with cte as (from cities)
 from array_of_arrays_of_cte();

┌─────────────────────────────────────────────────────────┐
│                          json                           │
│                          json                           │
├─────────────────────────────────────────────────────────┤
│ [["Amsterdam", 1, 0], ["London", 2, 0], ["Rome", 3, 0]] │
└─────────────────────────────────────────────────────────┘
```

**Comment by tonne1** [comment]
> I thought the obstacle you were seeking to overcome was how to avoid naming the columns.

Yes, that was the main issue, but also to get a JSON array of arrays.

> Agreed, this is cumbersome, to say the least. (Welcome to SQL!)

This is not due to SQL, in my view, I wouldn't know a more elegant and powerful yet readable language. This is due to the mapping between SQL objects and JSON (which IS complex) and the function support in duckdb so far.

Actually, you don't need a CTE, just use the original statement as derived table:

```sql
select array_agg(columns(*))::json 
from (select list_transform(json_keys(j), k -> (j -> k)) as json 
      from (select cities::json j from cities));
```

Still, what a counter-intuitive chunk of a statement just to get a JSON array of arrays from a simple query without enumerating the columns ;-)

A slightly more readable version could be

```sql
select array_agg(columns(*))::json 
from (select json_extract(j, json_keys(j)) as json 
      from (select cities::json j from cities));
```
but that gives: Binder Error: List of paths must be constant. 
Don't know why list_transform accepts json_keys() as argument and json_extract does not.

Thanks again for your input.

@duckdb team: Is there really no better way to achieve this?

**Reply by cmdlineluser** [reply]
It's similar to the request for `struct_keys()` - https://github.com/duckdb/duckdb/discussions/13454

There is no `struct_values()`

```python
>>> duckdb.sql("from cities select cities")
┌────────────────────────────────────┐
│               cities               │
│ struct("name" varchar, id integer) │
├────────────────────────────────────┤
│ {'name': Amsterdam, 'id': 1}       │
│ {'name': London, 'id': 2}          │
└────────────────────────────────────┘
```

```python
>>> duckdb.sql("from cities select list(cities)::json")
┌────────────────────────────────────────────────────────┐
│              CAST(list(cities) AS "json")              │
│                          json                          │
├────────────────────────────────────────────────────────┤
│ [{"name":"Amsterdam","id":1},{"name":"London","id":2}] │
└────────────────────────────────────────────────────────┘
```

If `struct_values` existed, presumably one would be able to do:

```sql
from cities select list(struct_values(cities))::json
```

**Comment by tonne1** [comment]
> 
```sql 
from cities select list(struct_values(cities))::json
```

@cmdlineluser thanks for the hint, that would be quite useful. I have added your idea to #13454.
